app/Pasteboard.m

```diff
@@ -0,0 +1,247 @@
+#include <string.h>
+#import <UIKit/UIKit.h>
+#include "fs/poll.h"
+#include "fs/dyndev.h"
+#include "kernel/errno.h"
+#include "debug.h"
+
+/**
+ * buffer is dynamically sized buffer of size buffer_cap
+ * All writes go to it, and buffer_len is length of data held in buffer
+ */
+
+// Prepare for fd separation
+#define fd_priv(fd) fd->clipboard
+typedef struct fd clip_fd;
+
+#define INITIAL_BUFFER_CAP 1024
+// 8MB: https://stackoverflow.com/a/3523175
+#define MAXIMAL_BUFFER_CAP 8*1024*1024
+
+// If pasteboard contents were changed since file was opened,
+// all read operations on in return error
+static int check_read_generation(clip_fd *fd) {
+    UIPasteboard *pb = UIPasteboard.generalPasteboard;
+
+    uint64_t pb_gen = (uint64_t) pb.changeCount;
+    uint64_t fd_gen = fd_priv(fd).generation;
+
+    if (fd_gen == 0 || fd->offset == 0) {
+        fd_priv(fd).generation = pb_gen;
+    } else if (fd_gen != pb_gen) {
+        return -1;
+    }
+
+    return 0;
+}
+
+static const char *get_data(clip_fd *fd, size_t *len) {
+    if (fd_priv(fd).buffer != NULL) {
+        *len = fd_priv(fd).buffer_len;
+        return fd_priv(fd).buffer;
+    }
+
+    if (check_read_generation(fd) != 0) {
+        return NULL;
+    }
+
+    NSString *contents = UIPasteboard.generalPasteboard.string;
+    *len = contents.length;
+    return contents.UTF8String;
```

Comment:
The UTF8String is freed when the NSString that it came from is freed, and that happens when this function returns. ASAN should warn you about this.

Reply:
I thought that variable was autoreleased for some reason.

Reply:
Looks like the default is __strong, though you could use __autoreleasing to autorelease instead

Reply:
Just checked -- unless I specifically wrap these 3 lines in `@autoreleasepool`, ASAN doesn't warn me about anything (because memory being referenced isn't actually freed).

Reply:
Although I should probably add autoreleasepool's wherever I use `get_data` to free that data when it's no longer needed.

Reply:
@tbodt sure, I'll add __autoreleasing, that makes sense.

Reply:
judging by disasm default seems to be strong indeed.